src/sync/remote_mongo_collection.cpp

```diff
@@ -18,3 +18,505 @@
+                                 args.dump(),
+                                 [completion_block](util::Optional<AppError> error, util::Optional<std::string> value) {
+            handle_response(error, value, completion_block);
+        });
+    } catch (const std::exception& e) {
+       return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), util::format("document parse %1", e.what())));
+    }
+}
+
+void RemoteMongoCollection::aggregate(std::vector<std::string> pipline,
+                                      std::function<void(std::string, util::Optional<AppError>)> completion_block)
+{
+    try {
+        auto pipelines = nlohmann::json::array();
+        for (std::string& pipeline_json : pipline) {
+            pipelines.push_back(nlohmann::json::parse(pipeline_json));
+        }
+        
+        auto base_args = m_base_operation_args;
+        base_args.push_back({ "pipeline", pipelines });
+        auto args = nlohmann::json( {{"arguments", nlohmann::json::array({base_args} ) }} );
+         
+        m_service->call_function("aggregate",
+                              args.dump(),
+                              [completion_block](util::Optional<AppError> error, util::Optional<std::string> value) {
+            handle_response(error, value, completion_block);
+        });
+    } catch (const std::exception& e) {
+        return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+    }
+}
+
+void RemoteMongoCollection::count(const std::string& filter_json,
+                                  uint64_t limit,
+                                  std::function<void(uint64_t, util::Optional<AppError>)> completion_block)
+{
+    try {
+        auto base_args = m_base_operation_args;
+        base_args.push_back({ "query", nlohmann::json::parse(filter_json) });
+        
+        if (limit != 0) {
+            base_args.push_back({ "limit", limit });
+        }
+        
+        auto args = nlohmann::json( {{"arguments", nlohmann::json::array({base_args} ) }} );
+
+        m_service->call_function("count",
+                                 args.dump(),
+                                 [completion_block](util::Optional<AppError> error, util::Optional<std::string> value) {
+            handle_count_response(error, value, completion_block);
+        });
+    } catch (const std::exception& e) {
+        return completion_block(0, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+    }
+}
+
+void RemoteMongoCollection::count(const std::string& filter_json,
+                                  std::function<void(uint64_t, util::Optional<AppError>)> completion_block)
+{
+    count(filter_json, 0, completion_block);
+}
+
+void RemoteMongoCollection::insert_many(std::vector<std::string> documents,
+                                        std::function<void(std::map<uint64_t, std::string>, util::Optional<AppError>)> completion_block)
```

Comment:
we're not really using the features of a map since everything is inserted in order, if there isn't a good reason for it, a std::vector will be more efficient

Reply:
I've moved to using a vector instead of map, it makes more sense